packages/labs/ssr/src/lib/module-loader.ts

```diff
@@ -0,0 +1,279 @@
+    const module = await modulePromise!;
+    return {
+      path: specifier,
+      module,
+    };
+  }
+
+  private _importModuleDynamically = async (
+    specifier: string,
+    referencingModule: vm.Module
+  ) => {
+    const result = await this.importModule(
+      specifier,
+      referencingModule.identifier
+    );
+    return result.module;
+  };
+
+  private _linker = async (
+    specifier: string,
+    referencingModule: vm.Module
+  ): Promise<vm.Module> => {
+    const {identifier} = referencingModule;
+    if (identifier.startsWith('file://')) {
+      if (!/:\d+$/.test(identifier)) {
+        throw new Error('Unexpected file:// URL identifier without context ID');
+      }
+      const referrer = identifier.split(/:\d+$/)[0];
+      return (await this._loadModule(specifier, referrer)).module;
+    } else {
+      throw new Error('unknown module source');
+    }
+  };
+
+  private _getIdentifier(modulePath: string) {
+    return 'file://' + modulePath + `:${this._vmContextId}`;
+  }
+
+  private _getBuiltInIdentifier(specifier: string) {
+    return specifier + `:${this._vmContextId}`;
+  }
+}
+
+/**
+ * Resolves specifiers using web-ish Node module resolution. Web-compatible
+ * full URLs are passed through unmodified. Relative and absolute URLs
+ * (starting in `/`, `./`, `../`) are resolved relative to `referrer`. "Bare"
+ * module specifiers are resolved with the 'resolve' package.
+ *
+ * This replaces some lit-html modules with SSR compatible equivalents. This is
+ * currently hard-coded, but should instead be done with a configuration object.
+ */
+export const resolveSpecifier = (specifier: string, referrer: string): URL => {
+  try {
+    // First see if the specifier is a full URL, and if so, use that.
+
+    // TODO: This will mainly be http:// and https:// URLs, which we may _not_
+    // want to support. We probably also want to filter out file:// URLs as
+    // those will be absolute to the file system.
+    return new URL(specifier);
+  } catch (e) {
+    if (referrer === undefined) {
+      throw new Error('referrer is undefined');
+    }
+    if (
+      specifierMatches(specifier, 'lit') ||
+      specifierMatches(specifier, 'lit-html') ||
+      specifierMatches(specifier, 'lit-element')
+    ) {
+      // Override where we resolve lit-html from so that we always resolve to
+      // a single version of lit-html.
+      referrer = import.meta.url;
+    }
+    const referencingModulePath = new URL(referrer).pathname;
+    const modulePath = resolve.sync(specifier, {
+      basedir: path.dirname(referencingModulePath),
+      moduleDirectory: ['node_modules'],
+      extensions: ['.js'],
+      // Some packages use a non-standard alternative to the "main" field
+      // in their package.json to differentiate their ES module version.
+      packageFilter: (packageJson: PackageJSON) => {
+        packageJson.main =
+          packageJson.module ?? packageJson['jsnext:main'] ?? packageJson.main!;
```

Comment:
I think `?? index.js` should be the final fallback?

Reply:
`resolve` should already implement that then. Here we're just adding in some non-standard behavior to look at the `module` and `jsnext:main` fields.